scipy/stats/_hypotests.py

```diff
@@ -1393,3 +1396,674 @@ def cramervonmises_2samp(x, y, method='auto'):
+                          "Consider reducing the batch size.")
+
+
+def permutation_test(data, statistic, *, permutation_type='both',
+                     vectorized=False, permutations=np.inf, batch=None,
+                     alternative="two-sided", axis=0, random_state=None):
+    r"""
+    Performs a permutation test of a given statistic on provided data.
+
+    For independent sample statistics, the null hypothesis is that the data are
+    randomly sampled from the same distribution.
+    For paired sample statistics, two null hypothesis can be tested:
+    that the data are paired at random or that the data are assigned to samples
+    at random.
+
+    Parameters
+    ----------
+    data : iterable of array-like
+        Contains the samples, each of which is an array of observations.
+        Dimensions of sample arrays must be compatible for broadcasting except
+        along `axis`.
+    statistic : callable
+        Statistic for which the p-value of the hypothesis test is to be
+        calculated. `statistic` must be a callable that accepts samples
+        as separate arguments (e.g. ``statistic(*data)``) and returns the
+        resulting statistic.
+        If `vectorized` is set ``True``, `statistic` must also accept a keyword
+        argument `axis` and be vectorized to compute the statistic along the
+        provided `axis` of the sample arrays.
+    permutation_type : {'both', 'sample', 'order'}
+        The type of permutations to be performed, in accordance with the
+        null hypothesis. The first two permutation types are for paired sample
+        statistics, in which all samples contain the same number of
+        observations and observations with corresponding indices along `axis`
+        are considered to be paired; the third is for independent sample
+        statistics.
+
+        - ``'samples'`` : observations are assigned to different samples
+          but remain paired with the same observations from other samples.
+          This permutation type is appropriate for paired sample hypothesis
+          tests such as the Wilcoxon signed-rank test and the paired t-test.
+        - ``'pairings'`` : observations are paired with different observations,
+          but they remain within the same sample. This permutation type is
+          appropriate for association/correlation tests with statistics such
+          as Spearman's :math:`\rho`, Kendall's :math:`\tau`, and Pearson's
+          :math:`r`.
+        - ``'both'`` : observations are assigned to different samples without
+          preserving pairs. Samples may contain different numbers of
+          observations. This permutation type is appropriate for
+          independent sample hypothesis tests such as the Mann-Whitney
+          :math:`U` test and the independent sample t-test.
+
+          Please see the Notes section below for more detailed descriptions
+          of the permutation types.
+
+    vectorized : bool, default: ``False``
```

Comment:
I think we normally say: `variable : type, optional`
and the default is specified in the description below that

Reply:
Sometimes, but this is allowed (see [numpydoc guidelines](https://numpydoc.readthedocs.io/en/latest/format.html#parameters)). The intent here (and e.g. [`bootstrap`](https://34267-1460385-gh.circle-artifacts.com/0/html-scipyorg/reference/generated/scipy.stats.bootstrap.html#scipy.stats.bootstrap)) is to include the default in the same line unless "the default value would not be used as a value", then I document it in the description.